update-vendor: include deps from all modules in root go.mod before copying

nikhita opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Currently, update-vendor:

**In Phase 3:**
a. Adds explicit `require` and `replace` directives for all dependencies in the root go.mod file.
b. Runs `go mod tidy` on the root repo
c. Propagates these directives to staging go.mod files

https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/hack/update-vendor.sh#L215-L224

**In Phase 4:**
a. Copies the root go.mod file to staging go.mod files

https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/hack/update-vendor.sh#L232-L235

**In Phase 5:**
a. Run `go mod tidy` on staging repos

https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/hack/update-vendor.sh#L314

If a staging go.mod file has an indirect dependency in its go.mod file but the root go.mod does not, step 4a ends up removing this dependency in the staging go.mod file.

Since this dependency does not exist in the staging go.mod file anymore, step 5a adds it back. However, `go mod tidy` will add a requirement on the latest version of each missing dep [1]. This means that the original pinned version in the staging go.mod file is not respected.

To fix this, this commit introduces the following changes:

1. Updates Phase 3 to:
a. include explicit `require` and `replace` directives for all dependencies in the root _and_ staging modules to create an aggregated go.mod file for the root go.mod.
b. propagate these to the staging go.mod files.
c. removes go mod tidy step here to ensure that we copy the complete root go.mod in Phase 4 to make sure we end up with consistent versions referenced.

2. Adds an additional step in Phase 5 to tidy the root go.mod file.

[1] https://go.dev/ref/mod#go-mod-tidy

#### Which issue(s) this PR fixes:

Fixes the root cause for https://github.com/kubernetes/kubernetes/issues/116946
xref https://github.com/kubernetes/kubernetes/pull/116948

#### Special notes for your reviewer:

`github.com/rogpeppe/go-internal` is an indirect dependency in some staging go.mod files, but is not in the root go.mod file. This triggers the above bug above where `go mod tidy` ended up adding a require directive to the latest version i.e. `v1.10.0` which was released recently - https://github.com/kubernetes/kubernetes/issues/116946#issuecomment-1485107586

Not sure if this change will have side-effects though, so:
/hold

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

/cc @liggitt @dims @MadhavJivrajani 
/assign @liggitt

MadhavJivrajani commented:
This is awesome, I'll take a deeper look but at the very least, we now know why the job started to fail, thank you @nikhita!

nikhita commented:
This will not work if we modify the root go.mod file. We'll need to the changes to get propagated to staging go.mod files too.

Will rework and send a fix.

nikhita commented:
> This will not work if we modify the root go.mod file. We'll need to the changes to get propagated to staging go.mod files too.

Correction: this works! I picked a random dependency `github.com/fvbommel/sortorder` and updated it to v1.1.0 and can confirm that this change is also propagated to staging go.mod repos.

Having said that, I had originally tried downgrading `golang.org/x/net` to v0.7.0. This didn't work. However, this doesn't seem to work on master either...

liggitt commented:
/hold
The copy from root to leaf is generally required to make sure we end up with consistent versions referenced

liggitt commented:
> If a staging go.mod file has an indirect dependency in its go.mod file but the root go.mod does not

I think this is the key thing we want to resolve. We want an aggregated go.mod that contains *all* the dependencies and versions from *all* our modules (root + staging)... that's what ensure_require_replace_directives_for_all_dependencies was trying to accomplish.

We then copy that to all the staging repos, and tidy all.

nikhita committed:
update-vendor: include deps from all modules in root go.mod before copying

Currently, update-vendor:

In Phase 3:
a. Adds explicit `require` and `replace` directives for all dependencies
in the root go.mod file.
b. Runs `go mod tidy` on the root repo
c. Propagates these directives to staging go.mod files

In Phase 4:
a. Copies the root go.mod file to staging go.mod files

In Phase 5:
a. Run `go mod tidy` on staging repos

If a staging go.mod file has an indirect dependency in
its go.mod file but the root go.mod does not, step 4a ends up
removing this dependency in the staging go.mod file.

Since this dependency does not exist in the staging go.mod file anymore,
step 5a re-adds it back. However, `go mod tidy` will add a requirement
on the latest version of each missing dep [1]. This means that the
original pinned version in the staging go.mod file is not respected.

To fix this, this commit introduces the following changes:

1. Updates Phase 3 to:
a. include explicit `require` and `replace` directives
for all dependencies in the root _and_ staging modules to create an
aggregated go.mod file for the root go.mod.
b. propagate these to the staging go.mod files.
c. removes go mod tidy step here to ensure that we copy the complete
root go.mod in Phase 4 to make sure we end up with consistent
versions referenced.

2. Adds an additional step in Phase 5 to tidy the root go.mod file.

[1] https://go.dev/ref/mod#go-mod-tidy

nikhita committed:
staging: update github.com/rogpeppe/go-internal in go.mod files

nikhita commented on staging/src/k8s.io/apimachinery/go.mod:
```diff
@@ -46,6 +46,7 @@ require (
 	github.com/onsi/gomega v1.27.4 // indirect
 	github.com/pkg/errors v0.9.1 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/rogpeppe/go-internal v1.10.0 // indirect
```
This change add go-internal to apimachinery, code-generator, etc and removed it from api, kube-proxy, etc.

Some more context:

```sh
# for apimachinery
$ go mod why -m github.com/rogpeppe/go-internal
# github.com/rogpeppe/go-internal
k8s.io/apimachinery/pkg/runtime
sigs.k8s.io/structured-merge-diff/v4/value
gopkg.in/yaml.v2
gopkg.in/yaml.v2.test
gopkg.in/check.v1
github.com/kr/pretty
github.com/rogpeppe/go-internal/fmtsort

# for api
$ go mod why -m github.com/rogpeppe/go-internal
# github.com/rogpeppe/go-internal
k8s.io/api/admission/v1
k8s.io/apimachinery/pkg/runtime
sigs.k8s.io/structured-merge-diff/v4/value
gopkg.in/yaml.v2
gopkg.in/yaml.v2.test
gopkg.in/check.v1
github.com/kr/pretty
github.com/rogpeppe/go-internal/fmtsort
````

liggitt replied:
I need to set aside more time to reason through this... this is surprising to me... I would have expected `// indirect` entries to already exist for all the modules which had paths to this dependency, which included apimachinery before this PR.

liggitt commented on hack/update-vendor.sh:
```diff
@@ -72,28 +72,54 @@ function ensure_require_replace_directives_for_all_dependencies() {
 
-  # tidy to ensure require directives are added for indirect dependencies
-  go mod tidy
+  # Propagate all replace/require directives into root and staging modules.
```
I think the root → staging propagation needs to happen _before_ we aggregate staging `require` directives back in

liggitt commented on hack/update-vendor.sh:
```diff
@@ -72,28 +72,54 @@ function ensure_require_replace_directives_for_all_dependencies() {
-          | xargs -L 100 go mod edit -fmt
-      jq -r '"-replace \(.Old.Path)=\(.New.Path)@\(.New.Version)"' < "${replace_json}" \
-          | xargs -L 100 go mod edit -fmt
+    collect_require_replace_directives "${require_json}" "${replace_json}"
```
we don't want to collect `replace` directives from staging repos (they shouldn't have any that differ from the root anyway)

maybe split to `collect_require_directives` and `collect_replace_directives`

liggitt commented on hack/update-vendor.sh:
```diff
@@ -72,28 +72,54 @@ function ensure_require_replace_directives_for_all_dependencies() {
-  # Capture local require/replace directives before running any go commands that can modify the go.mod file
+  # Capture local require/replace directives from root and staging modules
+  # before running any go commands that can modify the go.mod file
   local require_json="${local_tmp_dir}/require.json"
   local replace_json="${local_tmp_dir}/replace.json"
```
if we're not truncating on first write (with `>` redirect), ensure these files don't exist before calling `collect_require_replace_directives` for the first time

liggitt commented on hack/update-vendor.sh:
```diff
@@ -216,10 +242,6 @@ kube::util::list_staging_repos \
-# resolves/expands references in the root go.mod (if needed)
-go mod tidy >>"${LOG_FILE}" 2>&1
-# pin expanded versions
-ensure_require_replace_directives_for_all_dependencies
```
remind me why we're dropping this bit? isn't dropping the expanding / pinning problematic?

liggitt commented on hack/update-vendor.sh:
```diff
@@ -216,10 +242,6 @@ kube::util::list_staging_repos \
-# pin expanded versions
-ensure_require_replace_directives_for_all_dependencies
 # group require/replace directives
 group_directives
```
we need to tidy before calling group_directives so any expanded/added require directives get grouped appropriately

the later tidy calls are only expected to *remove* things, but the first time we call it, we might be *adding* indirect deps

cici37 commented:
/remove-sig api-machinery

bart0sh commented:
/remove-sig node

dims commented:
@nikhita please reopen when you rebase. is that ok?

thanks,
Dims
